nemo/collections/tts/models/tacotron2.py

```diff
@@ -0,0 +1,189 @@
+        mask = mask.expand(mel_target.shape[1], mask.size(0), mask.size(1))
+        mask = mask.permute(1, 0, 2)
+        mel_out.data.masked_fill_(mask, self.pad_value)
+        mel_out_postnet.data.masked_fill_(mask, self.pad_value)
+        gate_out.data.masked_fill_(mask[:, 0, :], 1e3)
+
+        gate_out = gate_out.view(-1, 1)
+        mel_loss = nn.MSELoss()(mel_out, mel_target) + nn.MSELoss()(mel_out_postnet, mel_target)
+        gate_loss = nn.BCEWithLogitsLoss()(gate_out, gate_target)
+        return mel_loss + gate_loss
+
+    def forward(self, audio, audio_len, tokens, token_len):
+        spec, spec_len = self.audio_to_melspec_precessor(audio, audio_len)
+        token_embedding = self.text_embedding(tokens).transpose(1, 2)
+        encoder_embedding = self.encoder(token_embedding, token_len)
+        if self.training:
+            spec_dec, gate, alignments = self.decoder(encoder_embedding, spec, memory_lengths=token_len)
+        else:
+            spec_dec, gate, alignments, _ = self.decoder.infer(encoder_embedding, memory_lengths=token_len)
+        spec_postnet = self.postnet(spec_dec)
+
+        max_len = spec.shape[2]
+        gate_padded = torch.zeros(spec_len.shape[0], max_len)
+        gate_padded = gate_padded.type_as(gate)
+        for i, length in enumerate(spec_len):
+            gate_padded[i, length.data - 1 :] = 1
+
+        return spec_dec, spec_postnet, gate, spec, gate_padded, spec_len, alignments
+
+    def training_step(self, batch, batch_idx):
+        audio, audio_len, tokens, token_len = batch
+        spec_dec, spec_postnet, gate, spec, gate_padded, spec_len, _ = self.forward(
+            audio, audio_len, tokens, token_len
+        )
+
+        loss = self.loss(
+            mel_out=spec_dec,
+            mel_out_postnet=spec_postnet,
+            gate_out=gate,
+            mel_target=spec,
+            gate_target=gate_padded,
+            target_len=spec_len,
+        )
+
+        output = {
+            'loss': loss,
+            'progress_bar': {'training_loss': loss},
+            'log': {'loss': loss},
+        }
+        return output
+
+    def validation_step(self, batch, batch_idx):
+        audio, audio_len, tokens, token_len = batch
+        spec_dec, spec_postnet, gate, spec, gate_padded, spec_len, alignments = self.forward(
+            audio, audio_len, tokens, token_len
+        )
+
+        loss = self.loss(
+            mel_out=spec_dec,
+            mel_out_postnet=spec_postnet,
+            gate_out=gate,
+            mel_target=spec,
+            gate_target=gate_padded,
+            target_len=spec_len,
+        )
+        return {
+            "loss": loss,
+            "mel_target": spec,
+            "mel_postnet": spec_postnet,
+            "gate": gate,
+            "gate_target": gate_padded,
+            "alignments": alignments,
+        }
+
+    def validation_epoch_end(self, outputs):
+        tacotron2_log_to_tb_func(
```

Comment:
why can't return log as part of result dictionary?

Reply:
with at least validation loss to log

Reply:
`tacotron2_log_to_tb_func` logs eval_loss by itself.

Reply:
Hmm I guess `ModelCheckpoint` needs something to be returned to be monitored...